File: backend/mcp-code-intelligence-nodejs/src/orchestration/registry/tokenizer.ts
/**
 * Text tokenizer for tool search — splits text into normalized tokens.
 * Handles: underscore_case, camelCase, hyphen-case, spaces.
 * Behavioral parity with Kotlin Tokenizer.kt.
 */

const STOPWORDS = new Set([
  'a', 'an', 'the', 'is', 'are', 'was', 'were', 'be', 'been',
  'to', 'for', 'and', 'or', 'in', 'on', 'with', 'from', 'by',
  'of', 'at', 'as', 'it', 'its', 'this', 'that', 'not', 'no',
]);

const SPLIT_RE = /[^a-zA-Z0-9]+/;
const CAMEL_RE = /(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])/;

/** Tokenize text into normalized, deduplicated, stopword-free tokens. */
export function tokenize(text: string): Set<string> {
  const rawParts = text.split(SPLIT_RE);
  const camelParts = text.split(CAMEL_RE);
  const allParts = [...rawParts, ...camelParts];
  const result = new Set<string>();
  for (const part of allParts) {
    const t = part.toLowerCase().trim();
    if (t.length > 1 && !STOPWORDS.has(t)) {
      result.add(t);
    }
  }
  return result;
}

/** Remove stopwords from a list of query terms. */
export function removeStopwords(terms: string[]): string[] {
  return terms.filter((t) => !STOPWORDS.has(t.toLowerCase()));
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/orchestration/registry/tokenizer.py
"""Text tokenizer for tool search — splits text into normalized tokens.

Handles: underscore_case, camelCase, hyphen-case, spaces.
Behavioral parity with Kotlin Tokenizer.kt.
"""

import re

STOPWORDS = frozenset({
    "a", "an", "the", "is", "are", "was", "were", "be", "been",
    "to", "for", "and", "or", "in", "on", "with", "from", "by",
    "of", "at", "as", "it", "its", "this", "that", "not", "no",
})

_SPLIT_RE = re.compile(r"[^a-zA-Z0-9]+")
_CAMEL_RE = re.compile(r"(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])")


def tokenize(text: str) -> set[str]:
    """Tokenize text into normalized, deduplicated, stopword-free tokens."""
    raw = _SPLIT_RE.split(text)
    camel_parts = _CAMEL_RE.split(text)
    all_parts = raw + camel_parts
    return {
        t
        for part in all_parts
        if (t := part.lower().strip()) and len(t) > 1 and t not in STOPWORDS
    }


def remove_stopwords(terms: list[str]) -> list[str]:
    """Remove stopwords from a list of query terms."""
    return [t for t in terms if t.lower() not in STOPWORDS]
